File: python/subarray-sum-equals-k.py
# -*- coding: utf-8 -*-
"""
Created on Tue Mar 10 11:48:26 2020

@author: shrey
"""

from collections import defaultdict 

class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        currSum = 0
        res = 0
        prevSum = defaultdict(lambda: 0)
        
        for num in nums:
            currSum += num
            
            if currSum == k:
                res += 1 
                
            if currSum - k  in prevSum:
                res += prevSum[currSum - k]
                
            prevSum[currSum] += 1
            
        return res

File: c++/subarray-sum-equals-k.cpp
#include <map>
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        map<int, int> prevSum;
        int res = 0;
        int currSum = 0;
        
        for(int i = 0; i< nums.size(); i++){
            currSum += nums[i];
            
            if(currSum == k){
                res++;
            }
            
            if(prevSum.find(currSum - k) != prevSum.end()){
                res += prevSum[currSum - k];
            }
            
            prevSum[currSum]++;
        }
        return res;
    }
};